### DISCUSSION TITLE
Nested Join not working on a many-to-many relationship

### DISCUSSION THREAD

**Discussion by shikhardadhich** [supabase/supabase#7519] [Questions] [answered]
I have 4 tables: 
1. Product_Master - Product_ID, Name 
2. Users - User_ID, Name
3. Invoice - Invoice_ID, User_ID (Foreign  key), User_Prodcut_List_ID (Foreign  key)
4. User_Product_List - User_Prodcut_List_ID, Product_ID(Foreign  key),  Invoice_ID ((Foreign  key), 

I am using following query and it works great, except I am not getting Product Name (getting only Ids to show in my Invoice page). I  am trying to generate invoice page with user detail and list  of product purchased by the user.  
 
**Generating Invoice:** 
supabase.from('Invoice').select(`
      *,
      User (*),
      User_Product_List(*)
      `)

Output - User, Invoice and Product ID 
Looking for - User, Invoice, Product Name 

**This I tried and it doesn’t work** 
supabase.from('Invoice').select(`
      *,
      User (*),
      User_Product_List(*),
      Product_Master (*)
      `),

All the foreign keys are correct, still its showing to references found in "Invoice" for Product_ID

**Comment by steve-chavez** [comment, marked as answer]
> supabase.from('Invoice').select(*, User (*), User_Product_List(*), Product_Master (*)),
> All the foreign keys are correct, still its showing to references found in "Invoice" for Product_ID

Is there an error message? Can you show a sample json output?

**Reply by shikhardadhich** [reply]
This is the error: 
_Could not find a relationship between 'Invoice' and 'Product_Master' in the schema cache_

**Reply by steve-chavez** [reply]
Ok. So there's no foreign key between `Invoice` and `Product_Master` thus no one-to-many or many-to-one relationship, also no junction table between those so no many-to-many relationship.

To do nesting you need relationships between the tables. How would you relate `Invoice` and `Product_Master` ?

**Reply by shikhardadhich** [reply]
Yes, there is no relation between Invoice and Product Master, however the table User_Product_List have Invoice_ID and Product_ID which makes a transitive dependence 
Invoice (_Invoice ID_) - > Product_List (_Invoice ID & Product ID_) -> Product Master (_Product ID)_

**Reply by steve-chavez** [reply]
I see, what's the `User_Product_List` primary key? If you include `Invoice_ID` and `Product_ID` in the primary key then this table will be recognized as a junction and there will be a m2m relationship.

```
PRIMARY KEY("User_Prodcut_List_ID", "Product_ID", "Invoice_ID")
```

**Reply by shikhardadhich** [reply]
Awesome! Thanks a lot! it worked

**Reply by Splix1** [reply]
> 

How can you do this from the supabase UI? I have the same problem now, saying "user" and "languages" have no relation and we have a "userLanguages" table with a tableId (the primary key), a userId foreign key and a languageId foreign key.

**Reply by steve-chavez** [reply]
@Splix1 On the Table Editor, you have the tables on the left panel, clicking the dropdown button should give you the "Edit Table" option. From there you can click the "Primary" checkbox to add columns to the PK.

**Reply by GaryAustin1** [reply]
@steve-chavez, Just ran into this helping a Discord user.  Spent awhile before finding this need for both foreign keys to be part of the primary key for a join table to work.  Don't think that is documented, but have not looked everywhere.  

Is this a newer requirement?  I don't remember needing to do that in a test a few months ago, but that code is long gone and may have been for something slightly different.  In this case neither of us knew to do that, until I searched here to see if there might be a bug or something we were doing wrong in the call.

**Reply by steve-chavez** [reply]
> Is this a newer requirement? 

@GaryAustin1 Yes, this is newish, it has been rolled out to Supabase using the latest postgrest pre-release.

It was done in: https://github.com/PostgREST/postgrest/pull/2262

The problem with the old many-to-many detection is that it was too liberal(it detected m2m relationships where the user didn't want it) and it generated errors like this one: https://github.com/supabase/supabase/discussions/6848.
